File: estudos/exercicios/java/Question34.java
package estudos.exercicios.java;

import java.util.concurrent.Semaphore;

public class Question34 {
    static int cont = 0;
    static Semaphore sem = new Semaphore(1);

    public static void main(String[] args) {
        Runnable a = (() -> {
            try {
                threadA();
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        });

        Runnable b = (() -> {
            try {
                threadB();
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        });

        Thread threadA = new Thread(a);
        Thread threadB = new Thread(b);

        threadB.start();
        threadA.start();

        try {
            threadA.join();
            threadB.join();
        } catch (Exception e) {
            Thread.currentThread().interrupt();
        }

        System.out.println(cont);

    }

    public static void threadA() throws InterruptedException {
        sem.acquire();
        cont++;
        sem.release();

    }

    public static void threadB() throws InterruptedException {
        sem.acquire();
        cont++;
        sem.release();
    }
}

File: estudos/exercicios/python/Question34.py
import threading

sem = threading.Semaphore(1)

count = 0


def ThreadA():
    global count
    sem.acquire()
    count = count + 1
    print(count)
    sem.release()


threads = []
for _ in range(10):
    thread = threading.Thread(target=ThreadA)
    threads.append(thread)
    thread.start()

for thread in threads:
    thread.join()

print("Final count:", count)
